flake-module.nix

```diff
@@ -21,7 +21,15 @@ in
             used by the `nix fmt` command.
           '';
           type = types.submoduleWith {
-            modules = (import ./.).all-modules pkgs ++ [{
+            modules = (import ./.).all-modules config.treefmt.pkgs ++ [{
+              options.pkgs = mkOption {
+                type = types.uniq (types.lazyAttrsOf (types.raw or types.unspecified));
+                description = ''
+                  Nixpkgs to use in `treefmt`.
+                '';
+                default = pkgs;
+                defaultText = "`pkgs` (module argument)";
+              };
```

Comment:
Although it _can_ all be done here, I believe it is better to make `default.nix` more aware of the use case of being loaded into a submodule. It doesn't have to know about flake-parts; what I'm suggesting is just to make this use case more visible by adding a `submodule-modules` module.

```suggestion
            modules = (import ./.).submodule-modules ++ [{
              options.pkgs = mkOption {
                defaultText = "`pkgs` (`perSystem` module argument)";
              };
```

`default.nix`
```nix
all-modules = ...;

# treefmt-nix can be loaded into a submodule. In this case we get our `pkgs` from
# our own standard option `pkgs`; not externally.
submodule-modules = 
   [
    {
      options.pkgs = mkOption {
        type = types.uniq (types.lazyAttrsOf (types.raw or types.unspecified));
        description = ''
          Nixpkgs to use in `treefmt`.
         '';
        default = pkgs;
      };
      config._module.args = {
        pkgs = nixpkgs;
      };
    }
    module-options
  ]
  ++ programs.modules
```

`_module.args.lib` doesn't do anything, because `lib` already behaves like a `specialArgs` argument, so that it can be used when computing `imports`.

Reply:
It's really just a refactor though; it behaves exactly the same, so let's do whichever solution the maintainers prefer.

Reply:
That sounds good. @shivaraj-bh do you mind integrating Robert's changes?

Reply:
Will do!

Reply:
Done here: https://github.com/numtide/treefmt-nix/pull/186/commits/3123e3763be68eee59dbc106a8b8c31bb73e06ec